File: cookbook/04-Feature/DebugTensor/main.py
# Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
# All rights reserved.
#
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import numpy as np
import tensorrt as trt
from tensorrt_cookbook import CookbookDebugListener, TRTWrapperV1

data = {"inputT0": np.arange(3 * 4 * 5, dtype=np.float32).reshape(3, 4, 5)}
expect_result = {"a_cute_tensor": data["inputT0"] * 3}  # the actual expected result is data*2, we set a wrong value here

tw = TRTWrapperV1()

tensor = tw.network.add_input("inputT0", trt.float32, data["inputT0"].shape)
layer1 = tw.network.add_elementwise(tensor, tensor, trt.ElementWiseOperation.SUM)
tensor1 = layer1.get_output(0)
tensor1.name = "a_cute_tensor"
layer2 = tw.network.add_elementwise(tensor1, tensor1, trt.ElementWiseOperation.SUM)

tw.network.mark_debug(tensor1)  # mark a tensor as debug-able
#tw.network.unmark_debug(tensor1)  # unmark a tensor as debug-able
print(f"{tw.network.is_debug_tensor(tensor1) = }")  # ensure one tensor is marked as debug-able

tw.build([layer2.get_output(0)])

tw.setup(data)

my_debug_listener = CookbookDebugListener(expect_result)
tw.context.set_debug_listener(my_debug_listener)  # set a debug listener for context
#debug_listener = context.get_debug_listener(CookbookDebugListener(expect_result))  # get a debug listener from context

tw.context.set_tensor_debug_state("a_cute_tensor", True)  # enable one debug tensor
#context.set_all_tensors_debug_state(True)  # enable all debug tensor

print(f"{tw.engine.is_debug_tensor('a_cute_tensor') = }")  # ensure one tensor is debug-able
print(f"{tw.context.get_debug_state('a_cute_tensor') = }")  # ensure one debug tensor is enabled

tw.infer()

# ------------------------------------------------------------------------------
# Usage of `mark_unfused_tensors_as_debug_tensors` + `unfused_tensors_debug_state`
# ------------------------------------------------------------------------------
# `mark_debug` prevents the marked tensor from being fused, which can change the
# optimized graph and hurt performance. `mark_unfused_tensors_as_debug_tensors`
# instead marks every tensor that survives fusion as a debug tensor, so it does
# not disturb fusion decisions (performance is preserved). The trade-off: these
# tensors are reported to the DebugListener by their internal (post-fusion) names
# rather than the names used in the NetworkDefinition, and they cannot be queried
# with `is_debug_tensor`.
tw2 = TRTWrapperV1()

tensor = tw2.network.add_input("inputT0", trt.float32, data["inputT0"].shape)
layer1 = tw2.network.add_elementwise(tensor, tensor, trt.ElementWiseOperation.SUM)
layer2 = tw2.network.add_elementwise(layer1.get_output(0), layer1.get_output(0), trt.ElementWiseOperation.SUM)

print(f"{tw2.network.mark_unfused_tensors_as_debug_tensors() = }")  # mark all unfused tensors as debug tensors
#tw2.network.unmark_unfused_tensors_as_debug_tensors()  # reverse operation

tw2.build([layer2.get_output(0)])
tw2.setup(data)

tw2.context.set_debug_listener(CookbookDebugListener())  # internal names are unknown here, so we just print them

tw2.context.unfused_tensors_debug_state = True  # enable emitting of all unfused debug tensors at runtime
print(f"{tw2.context.unfused_tensors_debug_state = }")  # read back the state

tw2.infer()

File: cookbook/04-Feature/DebugTensor/C++/main.cpp
/*
 * Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
 * All rights reserved.
 *
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// Two C++-side helpers from cookbookHelper, exercised on one small network:
//   + `DebugTensorWriter` - an IDebugListener that saves each debug tensor as a .npy file, so the
//     intermediate values can be diffed against a reference run in numpy afterwards. The Python
//     `CookbookDebugListener` only prints them.
//   + `launchGlobalTimerKernel` - reads the PTX %globaltimer register, the documented substitute
//     for cudaEventElapsedTime() under Confidential Compute.

#include "cookbookHelper.cuh"

using namespace nvinfer1;

// An IDebugListener that saves each debug tensor as a .npy file (through the bundled cnpy) instead
// of only printing it, so the values can be diffed against a reference run in numpy afterwards.
//
// This lives in the example rather than in cookbookHelper on purpose: it pulls in cnpy and zlib,
// and 30 other C++ examples link cookbookHelper without either of them.
class DebugTensorWriter : public IDebugListener
{
public:
    DebugTensorWriter(std::string prefix = "debug-tensor"):
        mPrefix(std::move(prefix)) {}

    bool processDebugTensor(void const *addr, TensorLocation location, DataType type, Dims const &shape, char const *name, cudaStream_t stream) noexcept override
    {
        int64_t nElement = std::accumulate(shape.d, shape.d + shape.nbDims, 1LL, std::multiplies<int64_t>());
        size_t  nByte    = nElement * dataTypeToSize(type);

        std::vector<char> hostBuffer(nByte);
        if (location == TensorLocation::kDEVICE)
        {
            CHECK(cudaStreamSynchronize(stream));
            CHECK(cudaMemcpyAsync(hostBuffer.data(), addr, nByte, cudaMemcpyDeviceToHost, stream));
            CHECK(cudaStreamSynchronize(stream));
        }
        else
        {
            std::memcpy(hostBuffer.data(), addr, nByte);
        }

        std::vector<size_t> cnpyShape;
        for (int i = 0; i < shape.nbDims; ++i)
        {
            cnpyShape.push_back(static_cast<size_t>(shape.d[i]));
        }

        // A tensor name can contain characters that are awkward in a path.
        std::string safeName(name);
        for (char &c : safeName)
        {
            if (c == '/' || c == '\\' || c == ' ' || c == ':')
            {
                c = '_';
            }
        }
        std::string fileName = mPrefix + "-" + safeName + ".npy";

        // cnpy needs the element type at compile time, so dispatch on the runtime DataType.
        switch (type)
        {
        case DataType::kFLOAT:
            cnpy::npy_save(fileName, reinterpret_cast<float const *>(hostBuffer.data()), cnpyShape, "w");
            break;
        case DataType::kINT32:
            cnpy::npy_save(fileName, reinterpret_cast<int32_t const *>(hostBuffer.data()), cnpyShape, "w");
            break;
        case DataType::kINT64:
            cnpy::npy_save(fileName, reinterpret_cast<int64_t const *>(hostBuffer.data()), cnpyShape, "w");
            break;
        default:
            // Types numpy has no direct equivalent for (FP16/BF16/FP8/INT4 ...) are written as raw
            // bytes, so they can still be reinterpreted on the numpy side.
            cnpy::npy_save(fileName, reinterpret_cast<uint8_t const *>(hostBuffer.data()), std::vector<size_t> {nByte}, "w");
            break;
        }

        std::cout << "[DebugTensorWriter] " << name << ", " << dataTypeToString(type) << ", "
                  << shapeToString(const_cast<Dims64 &>(shape)) << ", " << nByte << " bytes -> " << fileName << std::endl;
        return true;
    }

private:
    std::string mPrefix;
};

char const   *inputTensorName {"inputT0"};
char const   *debugTensorName {"a_cute_tensor"};
Dims64        shape {3, {3, 4, 5}};
static Logger gLogger(ILogger::Severity::kERROR);

int main()
{
    CHECK(cudaSetDevice(0));

    IBuilder           *builder = createInferBuilder(gLogger);
    INetworkDefinition *network = builder->createNetworkV2(0);
    IBuilderConfig     *config  = builder->createBuilderConfig();

    ITensor           *inputTensor = network->addInput(inputTensorName, DataType::kFLOAT, shape);
    IElementWiseLayer *layer1      = network->addElementWise(*inputTensor, *inputTensor, ElementWiseOperation::kSUM);
    ITensor           *tensor1     = layer1->getOutput(0);
    tensor1->setName(debugTensorName);
    // Marking a tensor as debuggable keeps it from being fused away, so it stays observable.
    network->markDebug(*tensor1);
    IElementWiseLayer *layer2 = network->addElementWise(*tensor1, *tensor1, ElementWiseOperation::kSUM);
    network->markOutput(*layer2->getOutput(0));

    IHostMemory *engineString = builder->buildSerializedNetwork(*network, *config);
    if (engineString == nullptr || engineString->size() == 0)
    {
        std::cout << "Fail building engine" << std::endl;
        return 1;
    }

    IRuntime          *runtime = createInferRuntime(gLogger);
    ICudaEngine       *engine  = runtime->deserializeCudaEngine(engineString->data(), engineString->size());
    IExecutionContext *context = engine->createExecutionContext();

    // Route debug tensors into .npy files instead of just printing them.
    DebugTensorWriter debugTensorWriter("debug-tensor");
    context->setDebugListener(&debugTensorWriter);
    context->setTensorDebugState(debugTensorName, true);

    int const                                        nIO = engine->getNbIOTensors();
    std::map<std::string, std::pair<void *, void *>> bufferMap; // name -> (host, device)
    for (int i = 0; i < nIO; ++i)
    {
        char const *name         = engine->getIOTensorName(i);
        Dims64      dim          = context->getTensorShape(name);
        int         nByte        = std::accumulate(dim.d, dim.d + dim.nbDims, 1, std::multiplies<>()) * dataTypeToSize(engine->getTensorDataType(name));
        void       *hostBuffer   = new char[nByte];
        void       *deviceBuffer = nullptr;
        CHECK(cudaMalloc(&deviceBuffer, nByte));
        bufferMap[name] = std::make_pair(hostBuffer, deviceBuffer);
        context->setTensorAddress(name, deviceBuffer);
    }

    float *pInput = static_cast<float *>(bufferMap[inputTensorName].first);
    for (int i = 0; i < shape.d[0] * shape.d[1] * shape.d[2]; ++i)
    {
        pInput[i] = float(i);
    }

    cudaStream_t stream;
    CHECK(cudaStreamCreate(&stream));

    int const nByteInput = shape.d[0] * shape.d[1] * shape.d[2] * sizeof(float);
    CHECK(cudaMemcpyAsync(bufferMap[inputTensorName].second, pInput, nByteInput, cudaMemcpyHostToDevice, stream));

    // Time the inference twice over: once with CUDA events, once with the GPU global timer. The two
    // should agree closely; under Confidential Compute only the second one stays trustworthy.
    uint64_t *dTimestamp = nullptr;
    CHECK(cudaMalloc(reinterpret_cast<void **>(&dTimestamp), 2 * sizeof(uint64_t)));

    cudaEvent_t eventStart, eventEnd;
    CHECK(cudaEventCreate(&eventStart));
    CHECK(cudaEventCreate(&eventEnd));

    CHECK(cudaEventRecord(eventStart, stream));
    CHECK(launchGlobalTimerKernel(dTimestamp, stream));

    context->enqueueV3(stream);

    CHECK(launchGlobalTimerKernel(dTimestamp + 1, stream));
    CHECK(cudaEventRecord(eventEnd, stream));
    CHECK(cudaStreamSynchronize(stream));

    float eventMs {0.0F};
    CHECK(cudaEventElapsedTime(&eventMs, eventStart, eventEnd));

    uint64_t hTimestamp[2] {};
    CHECK(cudaMemcpy(hTimestamp, dTimestamp, 2 * sizeof(uint64_t), cudaMemcpyDeviceToHost));
    double globalTimerMs = static_cast<double>(hTimestamp[1] - hTimestamp[0]) / 1.0e6;

    std::cout << "cudaEventElapsedTime : " << eventMs << " ms" << std::endl;
    std::cout << "%globaltimer         : " << globalTimerMs << " ms" << std::endl;

    CHECK(cudaEventDestroy(eventStart));
    CHECK(cudaEventDestroy(eventEnd));
    CHECK(cudaFree(dTimestamp));
    CHECK(cudaStreamDestroy(stream));

    for (auto const &[name, buffers] : bufferMap)
    {
        delete[] static_cast<char *>(buffers.first);
        CHECK(cudaFree(buffers.second));
    }
    delete context;
    delete engine;
    delete runtime;
    delete engineString;
    delete config;
    delete network;
    delete builder;

    std::cout << "Finish" << std::endl;
    return 0;
}
